storage/indexer/src/event_v2_translator.rs

```diff
@@ -0,0 +1,1327 @@
+            (key, sequence_number)
+        } else {
+            (EventKey::new(2, *coin_deposit.account()), 0)
+        };
+        let deposit_event = DepositEvent::new(coin_deposit.amount());
+        Ok(ContractEventV1::new(
+            key,
+            sequence_number,
+            DEPOSIT_EVENT_TYPE.clone(),
+            bcs::to_bytes(&deposit_event)?,
+        ))
+    }
+}
+
+struct CoinWithdrawTranslator;
+impl EventV2Translator for CoinWithdrawTranslator {
+    fn translate_event_v2_to_v1(
+        &self,
+        v2: &ContractEventV2,
+        engine: &EventV2TranslationEngine,
+    ) -> Result<ContractEventV1> {
+        let coin_withdraw = CoinWithdraw::try_from_bytes(v2.event_data())?;
+        let struct_tag_str = format!("0x1::coin::CoinStore<{}>", coin_withdraw.coin_type());
+        let struct_tag = StructTag::from_str(&struct_tag_str)?;
+        let (key, sequence_number) = if let Some(state_value_bytes) =
+            engine.get_state_value_bytes_for_resource(coin_withdraw.account(), &struct_tag)?
+        {
+            // We can use `DummyCoinType` as it does not affect the correctness of deserialization.
+            let coin_store_resource: CoinStoreResource<DummyCoinType> =
+                bcs::from_bytes(&state_value_bytes)?;
+            let key = *coin_store_resource.withdraw_events().key();
+            let sequence_number = engine
+                .get_next_sequence_number(&key, coin_store_resource.withdraw_events().count())?;
+            (key, sequence_number)
+        } else {
+            (EventKey::new(3, *coin_withdraw.account()), 0)
+        };
+        let withdraw_event = WithdrawEvent::new(coin_withdraw.amount());
+        Ok(ContractEventV1::new(
+            key,
+            sequence_number,
+            WITHDRAW_EVENT_TYPE.clone(),
+            bcs::to_bytes(&withdraw_event)?,
+        ))
+    }
+}
+
+struct CoinRegisterTranslator;
+impl EventV2Translator for CoinRegisterTranslator {
+    fn translate_event_v2_to_v1(
+        &self,
+        v2: &ContractEventV2,
+        engine: &EventV2TranslationEngine,
+    ) -> Result<ContractEventV1> {
+        let coin_register = CoinRegister::try_from_bytes(v2.event_data())?;
+        let struct_tag_str = "0x1::account::Account".to_string();
+        let struct_tag = StructTag::from_str(&struct_tag_str)?;
+        let (key, sequence_number) = if let Some(state_value_bytes) =
+            engine.get_state_value_bytes_for_resource(coin_register.account(), &struct_tag)?
+        {
+            let account_resource: AccountResource = bcs::from_bytes(&state_value_bytes)?;
+            let key = *account_resource.coin_register_events().key();
+            let sequence_number = engine
+                .get_next_sequence_number(&key, account_resource.coin_register_events().count())?;
+            (key, sequence_number)
+        } else {
+            (EventKey::new(0, *coin_register.account()), 0)
```

Comment:
ditto

Reply:
It's the creation number of the event key for `CoinRegisterEvent`. It's deterministically 0.